(SERVER-860) Introduce ReentrantReadWriteLock into JRubyPool data structure

rlinehan opened the pull request:
Rename RegisteredLinkedBlockingDeque to JRubyPool and move the locking logic relating to atomic code deploys into this data structure, so that we have one encapsulated datastructure managing the JRuby pool.

Add tests for our required behavior of this encapsulated data structure.

I have tried to separate out the commits so that adding/moving of files is in a separate commit from modifying those commits and is hopefully easier to understand.

rlinehan commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -299,9 +300,10 @@
+        (.returnItem pool instance)))
+    ;; if we get here, it was from a borrow and we got a Retry, so we just
+    ;; return it to the pool.
+    (.returnItem (:pool instance) instance)))
```
Copying this comment thread over from #698 (https://github.com/puppetlabs/puppet-server/pull/698#discussion_r40011586) because that was closed.

Chris:
>I think it's probably functionally equivalent, but, technically I think we're putting a Pill back into the pool here, so I wonder if just in terms of semantics it would be better to call your `insertPill` method?

My response:
>I spent a while trying to puzzle over this, and I think that this needs to be a `returnItem`. I changed the `putFirst` in `flush-pool!` to `insertPill` (https://github.com/puppetlabs/puppet-server/pull/698/files#diff-af969cd26304257b62d766cb4c5325faR118), because there there isn't anything holding a read lock; we're just adding the `RetryPoisonPill` to the pool for the first time.

>Here, on the other hand, it's possible that a borrow could have gotten the retry pill (https://github.com/puppetlabs/puppet-server/blob/master/src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj#L106). In that case, since it did a borrow it holds a read lock, which we need to unlock, which is why `returnItem`.

@cprice404 does what I'm saying here make sense, and should I stick with using `returnItem` here?

cprice404 replied:
OK, yep, I think you're right.  Good catch.

cprice404 commented:
@camlow325 if you end up with any bandwidth to review this PR, we'd definitely appreciate your input.

KevinCorcoran commented on src/java/com/puppetlabs/puppetserver/pool/LockablePool.java:
```diff
@@ -0,0 +1,25 @@
+    public void lock() throws InterruptedException;
+
+    public void unlock() throws InterruptedException;
+}
```
very minor Java note: all of the `public` modifiers on these methods are redundant; methods in an interface are all always public.

KevinCorcoran replied:
It would also be nice to have javadocs on these methods.

KevinCorcoran commented on test/unit/puppetlabs/puppetserver/lockable_pool_test.clj:
```diff
@@ -0,0 +1,193 @@
+        lock-acquired? (promise)
+        unlock? (promise)]
+    (is (= 3 (count instances)))
+    (let [lock-thread (future (do (deliver future-started? true)
```
This `do` is unnecessary (and there are several more of these below, which are also unnecessary).

KevinCorcoran commented on test/unit/puppetlabs/puppetserver/lockable_pool_test.clj:
```diff
@@ -0,0 +1,193 @@
+                                        (.returnItem pool instance))))]
+        ;; this is racey, but the only ways i could think of to make it non-racey
+        ;; depended on knowledge of the implementation
+        (Thread/sleep 500)
```
@cprice404 - Do you remember what those other ways were?  It seems quite possible that it would preferable for this test to depend on the implementation if that enabled the elimination of this race condition.

(sorry, I realize this is asking a question about code you *didn't* write several weeks ago).

KevinCorcoran replied:
... especially since the test below is dependent on the current implementation of `JRubyPool` and specifically calls that out.

KevinCorcoran commented:
:see_no_evil:

rlinehan commented:
I've pushed up several commits addressing feedback; mostly cleanup things and adding javadocs. I think I've addressed all feedback except for dealing with the race condition in `pool-lock-reentrant-with-many-borrows-test`.

cprice404 commented on test/unit/puppetlabs/puppetserver/lockable_pool_test.clj:
```diff
@@ -0,0 +1,191 @@
+        (is (thrown? IllegalStateException (.borrowItem pool)))
+        ;; (let [instance (.borrowItem pool)]
+        ;;   (is (true? true))
+        ;;   (.returnItem pool instance))
```
So, this whole test is effectively commented out because we don't acutually support re-entrant borrows... so it seems like it would probably make sense to just comment out those lines as well?  And worry about race conditions when we move to an implementation that is re-entrant?

KevinCorcoran replied:
that would be fine with me

rlinehan replied:
addressed in 9eb12e1aea53421b6fa8a038ef7938449fc08f04

KevinCorcoran commented on src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java:
```diff
@@ -0,0 +1,194 @@
+    * to the queue and the list of registered instances are visible roughly
+    * atomically to consumers, but because the underlying queue uses its own
+    * lock, it is possible for it to be modified on another thread while this
+    * method is being executed.
```
I got pretty hung-up trying to understand this paragraph.  It particular, I think the phrase `visible roughly atomically` is pretty confusing.

rlinehan replied:
I just copied back in the paragraph that Chris had here before. TBH I also find it pretty hard to understand. I'll try to see if I can improve it though...

rlinehan replied:
In the process of trying to come up with a better description of this, I discovered that I messed up in a previous commit and accidentally reordered the `.putLast` and the `.add` to `registeredElements` (https://github.com/rlinehan/puppet-server/commit/67afeb70eb1bf813009bdf2a8d5c28137f135eb7#diff-42395ae4a364917085c26d57d9e4cf85R20). I've added a commit 8a8e636 fixing this. (Thanks @scottyw for helping me work through this!)

rlinehan replied:
Trying to come up with a better explanation of this...

~~~
This method is synchronized not for thread safety, but to try to ensure that
to consumers of this class adding an instance to the queue and adding it to
the set of registered elements is visible roughly atomically. `synchronize`
guarantees that registration is atomic, so (ignoring borrows happening)
`RegisteredElements` and `liveQueue` can only diverge by 1. "This is only
"roughly" atomic, as the underlying queue uses its own lock, so it is
possible for it to be modified on another thread while this method is being
executed.
 ~~~

suggestions/improvements welcome.

cprice404 replied:
That sounds fine (and better).  TBH I don't think it is worth spending a bunch of time on this right now so I'd say change it to that and then let's put this to bed.

KevinCorcoran commented on src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java:
```diff
@@ -0,0 +1,194 @@
+    /**
+    * Acquires a write lock from <tt>ReentrantReadWriteLock</tt>.
+    *
+    * Note that this implementation does not fulfill requirement (e); with
```
What is `requirement (e)`?

KevinCorcoran replied:
Oh, I see, it's in the javadocs on the interface.  Might want to explain that.

KevinCorcoran commented:
:+1:

cprice404 commented:
@camlow325 what's the likelihood you'd be able to give this a quick skim tomorrow?  I think we need to get it merged before EOW so if you won't have bandwidth, we'll just have to merge it anyway and maybe you can retroactively review later?

camlow325 commented:
@cprice404 Have a few things I want to take care of this morning but will aim to take a look at this today.  If I haven't gotten to it by then, though, I'm good with ye just merging it anyway, and I could then take a look back at it later.

cprice404 commented:
@camlow325 merging this as it is blocking other work.  Would still appreciate your input.
